docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+      * [As an infrastructure provider author, I would like to build a controller to manage provisioning machines without being restricted to a CRUD API\.](#as-an-infrastructure-provider-author-i-would-like-to-build-a-controller-to-manage-provisioning-machines-without-being-restricted-to-a-crud-api)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Machine Controller Role](#machine-controller-role)
+      * [Machine Controller dynamic watchers](#machine-controller-dynamic-watchers)
+      * [Object References, Templates, MachineSets and MachineDeployments](#object-references-templates-machinesets-and-machinedeployments)
+      * [Controllers and the single responsibility approach](#controllers-and-the-single-responsibility-approach)
+      * [Remote references and accessing a workload cluster](#remote-references-and-accessing-a-workload-cluster)
+      * [The “Phase” field and its role](#the-phase-field-and-its-role)
+      * [Showing a status summary to users](#showing-a-status-summary-to-users)
+    * [Risks and Mitigations](#risks-and-mitigations)
+      * [State transitions are inflexible](#state-transitions-are-inflexible)
+      * [Machine Controller can access any machine or cluster in any namespace](#machine-controller-can-access-any-machine-or-cluster-in-any-namespace)
+      * [Certificates and tokens are exposed in plaintext](#certificates-and-tokens-are-exposed-in-plaintext)
+      * [Bootstrap data cannot be merged](#bootstrap-data-cannot-be-merged)
+      * [MachineClass is deprecated and will be revisited later](#machineclass-is-deprecated-and-will-be-revisited-later)
+  * [Design Details](#design-details)
+    * [Test Plan](#test-plan)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    * [Version Skew Strategy](#version-skew-strategy)
+  * [Implementation History](#implementation-history)
+  * [Drawbacks](#drawbacks)
+  * [Alternatives](#alternatives)
+      * [Object References, Templates, MachineSets and MachineDeployments](#object-references-templates-machinesets-and-machinedeployments-1)
+
+## Glossary
+
+- **Cluster API**: Unless otherwise specified, this refers to the project as a whole.
+- **Machine**: the Custom Resource for Kubernetes
+- **Server/Instance/Host**: the infrastructure that backs a Machine
+- **Bootstrapping**: turning a server into a Kubernetes node
+
+## Summary
+
+This proposal outlines splitting up the Machine data model into two new providers/controllers along with the generic Machine controller. The two new providers are the Infrastructure provider and the Bootstrap provider. The infrastructure provider handles provisioning infrastructure (cloud, bare metal, virtual) and the bootstrap provider handles Kubernetes bootstrapping. This change will improve separation of concerns for the Machine controller and gives more degrees of freedom to the user.
```

Comment:
I like the bootstrap defined in terms of "turning a <server> into a kubernetes node", but I'm not so fond about using "server", mostly because the exact meaning of a "machine" is provider dependent. I understand here "infrastructure" refers to "execution platform and related resources like network and storage".

Reply:
Thanks for the feedback! I incorporated some of the feedback you suggested. I've checked through the document and we don't refer to "Infrastructure provider" with a capitalized word. In this case, "infrastructure provider" or "bootstrap provider" are referred to as types of a provider.